### PR TITLE
Validate blob issued to staging (must match app)

### LINKED ISSUE #801: Validate blob belongs to the app when staging
Depends on #699

When the user uploads the application code, we [store the application metadata](https://github.com/epinio/epinio/pull/788/files#diff-4c92e03256b6a12e87b841840188789dbcf00fcb04087875af0779129ac3cf3bR79-R81) on the object itself. We do the same [when we pull from git](https://github.com/epinio/epinio/pull/788/files#diff-676e4bf1f445a3687fbd3165180a5d030da0b4438d7a5ae481f67ebb758e4054R83-R85)

We can use these metadata to validate that the blob the client sends on the stage request actually matches the application on the same request. This should prevent someone from staging an app with the sources of another app.

Given everyone can do everything on Epinio right now (we don't have user roles), this is mostly for validation against code mistakes on the client (e.g. the web UI).

After discussing it with @manno, we think it could even be a feature to be able to deploy applications with the blob of another app. E.g. Someone updates the sources of a wordpress application and everyone else is staging their apps using those (updated) sources. To make it a useful feature, it would require a lot more that simply allowing this.

TBD: To validate or not to validate?

### PR DESCRIPTION
**Pull request by andreas-kupries**
Fixes #801

feat: extended s3 manager with method to retrieve user meta data
feat: extended stage endpoint with validation of blob meta data

### CODE REVIEW

File: internal/s3manager/s3manager.go [hunk 1]
```diff
@@ -158,6 +158,21 @@ region = %s
 	return secret, err
 }
 
+// Stat retrieves the meta data for the blob specified by it blobUID.
```
Comment (jimmykarily) [inline internal/s3manager/s3manager.go:161]:
```suggestion
// Meta retrieves the meta data for the blob specified by it blobUID.
```

File: internal/s3manager/s3manager.go [hunk 2]
```diff
@@ -158,6 +158,21 @@ region = %s
...
+func (m *Manager) Meta(ctx context.Context, blobUID string) (map[string]string, error) {
+	if err := m.EnsureBucket(ctx); err != nil {
```
Comment (jimmykarily) [inline internal/s3manager/s3manager.go:163]: This line will make a request to S3 every time. If the bucket is not there yet, it will create it but this method will fail later because obviously the object doesn't exist (since the bucket didn't exist either). We could as well let it fail because of the missing bucket. After all, it's a bit unexpected for this "getter" method to create the bucket. One can argue that it's also the case for the `Upload` method which does the same but at least there, the creation of the bucket is happening so that the method doesn't fail.

In other words, we can remove this call to `EnsureBucket` here.

File: internal/api/v1/application/stage.go [hunk 3]
```diff
@@ -134,6 +134,48 @@ func (hc Controller) Stage(c *gin.Context) apierror.APIErrors {
...
 	}
 
+	// Validate incoming blob id before attempting to stage
+
+	s3ConnectionDetails, err := s3manager.GetConnectionDetails(ctx, cluster, deployments.TektonStagingNamespace, deployments.S3ConnectionDetailsSecret)
+	if err != nil {
+		return apierror.InternalError(err, "failed to fetch the S3 connection details")
+	}
+
+	manager, err := s3manager.New(s3ConnectionDetails)
```
Comment (jimmykarily) [inline internal/api/v1/application/stage.go:144]: All the additions below this point can be extracted to a function that does what the comment above says. E.g.

```
validateBlob(req.BlobUID, req.App, s3ConnectionDetails)
```

**Review by jimmykarily — CHANGES_REQUESTED** [review]
Can we add a test for this? Something like:

```
Create app1 and upload its code
Create app2 and upload its code
Try to stage app2 with the blobUID of app1
Expect failure
```

File: internal/s3manager/s3manager.go (hunk 2 above)
Comment (andreas-kupries) [reply internal/s3manager/s3manager.go:163]: Ok. Will do. Will also look into writing a test.

File: internal/api/v1/application/stage.go (hunk 3 above)
Comment (andreas-kupries) [reply internal/api/v1/application/stage.go:144]: Will do.

Comment (andreas-kupries) [reply internal/api/v1/application/stage.go]: commit 189a36981220ee32960d620612dadaf51204995d

File: internal/s3manager/s3manager.go (hunk 2 above)
Comment (andreas-kupries) [reply internal/s3manager/s3manager.go]: commit 189a36981220ee32960d620612dadaf51204995d

todo: test

Comment (andreas-kupries) [reply internal/s3manager/s3manager.go]: test commit 96a76c10c5f8d2ca795b8463549f51fb92bb6d78

**Comment by andreas-kupries** [comment]
The failed test is some cert issue in the installation

**Review by jimmykarily — APPROVED** [review]
